File: Sources/Olaf/UI/Views/Badges.swift
#if canImport(UIKit)
import SwiftUI

/// Colored pill for the HTTP status code / error.
struct StatusPill: View {
    let statusCode: Int?
    let isFailure: Bool

    var body: some View {
        Text(text)
            .font(.caption2.weight(.bold).monospacedDigit())
            .foregroundStyle(.white)
            .padding(.horizontal, 7)
            .padding(.vertical, 3)
            .background(color, in: RoundedRectangle(cornerRadius: 6, style: .continuous))
    }

    private var text: String {
        if let statusCode { return String(statusCode) }
        return isFailure ? "ERR" : "•••"
    }

    private var color: Color {
        if isFailure { return .red }
        guard let statusCode else { return .gray }
        switch statusCode {
        case 200..<300: return .green
        case 300..<400: return .teal
        case 400..<500: return .orange
        default: return .red
        }
    }
}

/// HTTP method badge (monospace).
struct MethodBadge: View {
    let method: String

    var body: some View {
        Text(method.uppercased())
            .font(.caption2.weight(.semibold).monospaced())
            .foregroundStyle(.secondary)
            .padding(.horizontal, 5)
            .padding(.vertical, 2)
            .background(Color.secondary.opacity(0.15), in: RoundedRectangle(cornerRadius: 4, style: .continuous))
    }
}

/// Colored dot on the left for the log level.
struct LevelDot: View {
    let level: LogLevel
    var body: some View {
        Circle()
            .fill(level.color)
            .frame(width: 8, height: 8)
    }
}
#endif

File: Android/olaf/src/main/java/com/olaf/ui/view/Badges.kt
package com.olaf.ui.view

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.olaf.LogLevel

/** Coloured pill carrying the HTTP status code, or the failure marker. */
@Composable
internal fun StatusPill(statusCode: Int?, isFailure: Boolean, modifier: Modifier = Modifier) {
    val text = statusCode?.toString() ?: if (isFailure) "ERR" else "•••"
    Text(
        text = text,
        color = Color.White,
        fontWeight = FontWeight.Bold,
        fontFamily = FontFamily.Monospace,
        style = MaterialTheme.typography.labelSmall,
        modifier = modifier
            .background(statusColor(statusCode, isFailure), RoundedCornerShape(6.dp))
            .padding(horizontal = 7.dp, vertical = 3.dp)
    )
}

/** HTTP method badge. */
@Composable
internal fun MethodBadge(method: String, modifier: Modifier = Modifier) {
    Text(
        text = method.uppercase(),
        color = MaterialTheme.colorScheme.onSurfaceVariant,
        fontWeight = FontWeight.SemiBold,
        fontFamily = FontFamily.Monospace,
        style = MaterialTheme.typography.labelSmall,
        modifier = modifier
            .background(MaterialTheme.colorScheme.surfaceVariant, RoundedCornerShape(4.dp))
            .padding(horizontal = 5.dp, vertical = 2.dp)
    )
}

/** Marks a mocked response, so a canned reply is never mistaken for a real one. */
@Composable
internal fun MockBadge(modifier: Modifier = Modifier) {
    Text(
        text = "MOCK",
        color = Color.White,
        fontWeight = FontWeight.Bold,
        style = MaterialTheme.typography.labelSmall,
        modifier = modifier
            .background(OlafColors.Teal, RoundedCornerShape(50))
            .padding(horizontal = 6.dp, vertical = 1.dp)
    )
}

/** The level dot shown at the leading edge of a plain log row. */
@Composable
internal fun LevelDot(level: LogLevel, modifier: Modifier = Modifier) {
    Box(
        modifier = modifier
            .size(8.dp)
            .background(level.color(), CircleShape)
    )
}
